routing/single_mwm_router.cpp

```diff
@@ -215,6 +197,46 @@ bool SingleMwmRouter::LoadIndex(MwmSet::MwmId const & mwmId, string const & coun
   }
 }
 
+void SingleMwmRouter::BuildRoute(MwmSet::MwmId const & mwmId, IndexGraphStarter & starter, vector<Joint::Id> const & joints, RouterDelegate const & delegate, Route & route) const
+{
+  vector<RoutePoint> routePoints;
+  starter.RedressRoute(joints, routePoints);
+
+  vector<Junction> junctions;
+  junctions.reserve(routePoints.size());
+
+  Geometry & geometry = starter.GetGraph().GetGeometry();
+  // TODO: Use real altitudes for pedestrian and bicycle routing.
+  for (RoutePoint const & routePoint : routePoints)
+    junctions.emplace_back(geometry.GetPoint(routePoint.GetRoadPoint()), feature::kDefaultAltitudeMeters);
+
+  shared_ptr<traffic::TrafficInfo::Coloring> trafficColoring = m_trafficCache.GetTrafficInfo(mwmId);
+  ReconstructRoute(m_directionsEngine.get(), m_roadGraph, trafficColoring, delegate, junctions, route);
+
+  // ReconstructRoute duplicates all points except start and finish.
+  // Therefore one needs fix time indexes to fit reconstructed polyline.
+  // TODO: rework ReconstructRoute and remove this stuff.
+  if (routePoints.size() >= 2 && route.GetPoly().GetSize() + 2 == routePoints.size() * 2)
```

Comment:
Feel free to return false if this condition is not satisfied, because as far as I know route.GetPoly().GetSize() + 2 should always be equal to routePoints.size() in the current implementation. So, print error message and return false. On the call place, check return status and return something appropriate.